### PR TITLE
Fix cpu compile

### PR DESCRIPTION
**Pull request by awni**
- Fix file name length bug for CPU compile
- Add test for another bug causing a segfault when all inputs are scalars

Simplified the contiguity stuff for the CPU it was too complicated and I think buggy, it just uses the exact same logic as the GPU eval now. We may want to improve both the CPU and GPU contiguity conditions at some point, but I think it needs more care.

Transformer training runs now with CPU compile.

### CODE REVIEW

**Review by angeloskath — APPROVED** [review]
Looks great!

**Comment by awni** [comment]
@angeloskath I changed this a little to keep the contiguity check more generous but refactor the logic to clean it up (it's shared between CPU and GPU now). I think it's right as it is...lmk if you want to look again.

File: mlx/backend/common/compiled.cpp [hunk 1]
```diff
@@ -126,4 +126,101 @@ std::string build_lib_name(
...
+    array::Flags flags;
+    for (int i = 0; i < inputs.size() && o < outputs.size(); ++i) {
+      auto& in = inputs[i];
+      // Conditions for donation
+      // - Contiguous
+      // - Donatable
+      // - Correct size
+      // - Not a constant
+      if (in.flags().contiguous && !is_scalar(in) && in.is_donatable() &&
+          constant_ids_.find(inputs_[i].id()) == constant_ids_.end()) {
```
Comment (angeloskath) [inline mlx/backend/common/compiled.cpp:177]: I think we still need to check `nbytes()` to make sure the types are the same or `itemsize()` to be more explicit. Also presumably the `in.flags().contiguous` check is not needed as it is covered by `is_scalar(in)`.

**Review by angeloskath — APPROVED** [review]
I left a comment for what I think is an edge-case. The refactoring was so needed, thanks!

**Pull request merged**